core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java

```diff
@@ -433,13 +434,30 @@ public String startWorkflow(
         }
     }
 
+    /*
+     * Acquire and hold the lock till the workflow creation action is completed (in primary and secondary datastores).
+     * This is to ensure that workflow creation action precedes any other action on a given workflow.
+     */
+    private void createWorkflow(Workflow workflow) {
+        if (!executionLockService.acquireLock(workflow.getWorkflowId())) {
+            throw new ApplicationException(BACKEND_ERROR, "Error acquiring lock when creating workflow: {}");
+        }
+        try {
+            executionDAOFacade.createWorkflow(workflow);
```

Comment:
It's tangential to this change, but i was reading the code and see that `ExecutionDAOFacade.removeWorkflow` does not remove the workflow id from the decider queue. Technically, it's possible that `IndexDAO.indexWorkflow` could fail and in that case, it's up to the `WorkflowSweeper` to remove it.

For symmetry, if `ExecutionDAOFacade.removeWorkflow` is supposed to undo `ExecutionDAOFacade.createWorkflow`, shouldn't it try to undo everything createWorkflow did?

Reply:
good point, added remove from decider queue to `removeWorkflow`.